streams/shared/src/main/scala/zio/stream/Structure.scala

```diff
@@ -0,0 +1,148 @@
+package zio.stream
+
+import zio._
+import ZStream.Pull
+
+private[stream] sealed abstract class Structure[-R, +E, +A] { self =>
+  import Structure._
+
+  def ++[R1 <: R, E1 >: E, A1 >: A](other: => Structure[R1, E1, A1]): Structure[R1, E1, A1] =
+    Concat(self, () => other)
+
+  def map[B](f: A => B): Structure[R, E, B] =
+    self match {
+      case Iterator(iter)          => Structure.Iterator(iter.map(_.map(f)))
+      case Concat(hd, tl)          => Structure.Concat(hd.map(f), () => tl().map(f))
+      case s: FlatMap[R, E, a0, A] => Structure.FlatMap(s.fa, (el: a0) => s.f(el).map(f))
+      case Effect(zio)             => Structure.Effect(zio.map(f))
+    }
+
+  def mapM[R1 <: R, E1 >: E, B](f: A => ZIO[R1, E1, B]): Structure[R1, E1, B] =
+    self match {
+      case Concat(hd, tl)          => Concat(hd.mapM(f), () => tl().mapM(f))
+      case Iterator(iter)          => Iterator(iter.map(_.flatMap(f(_).mapError(Some(_)))))
+      case s: FlatMap[R, E, a0, A] => Structure.FlatMap(s.fa, (el: a0) => s.f(el).mapM(f))
+      case Effect(zio)             => Structure.Effect(zio.flatMap(f))
+    }
+
+  def flatMap[R1 <: R, E1 >: E, B](f: A => Structure[R1, E1, B]): Structure[R1, E1, B] =
+    FlatMap[R1, E1, A, B](self, f)
+
+  def process: ZManaged[R, E, Pull[R, E, A]] =
+    self match {
+      case Iterator(iter) => iter
+      case _ =>
+        for {
+          refs <- (Ref.make[Structure[R, E, _]](self) <*>
+                   Ref.make[Pull[R, E, _]](Pull.end) <*>
+                   Ref.make[Exit[_, _] => URIO[R, _]](_ => UIO.unit) <*>
+                   Ref.make[List[(Pull[R, E, _], Exit[_, _] => URIO[R, _], Structure[R, E, _])]](Nil)).toManaged_
```

Comment:
Just as a side note, we now have so many of these `Ref.make(...).toManaged_` that we should add a `Ref.makeManaged`.